diff --git a/mlx/io/load.cpp b/mlx/io/load.cpp
index a9379c307..ebbb2a85d 100644
--- a/mlx/io/load.cpp
+++ b/mlx/io/load.cpp
@@ -4,6 +4,7 @@
 #include <fstream>
 #include <limits>
 #include <sstream>
+#include <thread>
 
 // Used by pread implementation.
 #ifdef _WIN32
@@ -334,9 +335,20 @@ array load(std::string file, StreamOrDevice s) {
 
 namespace io {
 
+size_t io_thread_count() {
+  constexpr size_t minimum_io_thread_count = 4;
+  constexpr size_t maximum_io_thread_count = 8;
+  const auto logical_cpu_count = std::thread::hardware_concurrency();
+  const auto half_logical_cpu_count = (logical_cpu_count + 1) / 2;
+  return std::clamp(
+      static_cast<size_t>(half_logical_cpu_count),
+      minimum_io_thread_count,
+      maximum_io_thread_count);
+}
+
 ThreadPool& thread_pool() {
   // Leak - see Scheduler singleton comment in scheduler.cpp.
-  static ThreadPool* pool_ = new ThreadPool{4};
+  static ThreadPool* pool_ = new ThreadPool{io_thread_count()};
   return *pool_;
 }
 
